### ISSUE TITLE
DontShow should not hide common parameters

### ISSUE THREAD

**Issue by MartinGC94** [PowerShell/PowerShell#18472]
### Summary of the new feature / enhancement

The DontShow attribute parameter that the Parameter attribute has allows a function/script author to hide parameters like this:

```
function Verb-Noun
{
    Param
    (
        [Parameter()]
        [string]
        $Param1,

        [Parameter(DontShow)]
        [string]
        $Param2
    )
}
#Tab completion does not show Param2
Verb-Noun -<Tab>
```

This has the side effect of also hiding the common parameters.  
It seems to have been an intentional design choice, see this: https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs#L768 but I think that should be changed because it makes it impractical to use `DontShow` to hide old deprecated parameters like `ConvertTo-Csv` does with the `NoTypeInformation` parameter because then the user won't see important common parameters like `ErrorAction`.

### Proposed technical implementation details (optional)

Instead of automatically hiding common parameters when `DontShow` has been specified, add a `HideCommonParameters` parameter to the `CmdletBinding` attribute.

**Comment by 237dmitry** [comment]
If you type the first letter of common parameter then `<TAB>` is working. Otherwise, how to understand which parameter you are going to enter? `-Param2` or for example `-ErrorAction`.

**Comment by kilasuit** [comment]
Alternatively a change on L780 from false to true would be the simplest way to solve this `This has the side effect of also hiding the common parameters.`

https://github.com/PowerShell/PowerShell/blob/9f459c05f617fe530dac340aa546d198724d85da/src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs#L772-L781

However expanding on your proposed implementation this I think that if a `HideCommonParameters` option to `CmdletBinding` was added then there would need to be a check for it.
My question would be `Is there really a need for us to hide the common parameters?` & if not then do we really need a `HideCommonParameters` option for the `CmdletBinding` attribute?

**Comment by MartinGC94** [comment]
I can see why someone might want to hide the 11 common parameters you get with cmdletbinding in simple functions/scripts that are meant to be used by less technical users. So I think it's best to have a replacement instead of just completely removing the option.

**Comment by kilasuit** [comment]
>I can see why someone might want to hide the 11 common parameters you get with cmdletbinding in simple functions/scripts that are meant to be used by less technical users.

You can do this by omitting the `[parameter()]` attribute off of all parameters entirely

``` powershell
function Verb-Nounsimple { 
    Param
    (
        [string]
        $Param1,

        [string]
        $Param2
    )
}
```

or also commonly written like so 

``` powershell
function Verb-Nounsimple { param([string]$Param1,[string]$Param2) }
```
wont give you the common parameters in tab completion

Same applies if you were to have a script with simple param block as per the above verb-nounsimple example

**Comment by MartinGC94** [comment]
Sure you can, but then you miss out on pipeline support, parametersets, and mandatory parameters.

**Comment by kilasuit** [comment]
> Sure you can, but then you miss out on pipeline support, parametersets, and mandatory parameters.

Then you want an advanced function/script and not a simple function / script which again comes back to a more important question on function/script design. which is a much larger question than many realise until later on in the development of it & often changes drastically over time.

**Comment by MartinGC94** [comment]
> > Sure you can, but then you miss out on pipeline support, parametersets, and mandatory parameters.
> 
> Then you want an advanced function/script and not a simple function / script which again comes back to a more important question on function/script design. which is a much larger question than many realise until later on in the development of it & often changes drastically over time.

So what's your point here? That nobody would want to hide the common parameters for a random script with a mandatory parameter?  
I'm sure I'm not the only one that has done something like this in a script where I needed a mandatory parameter but didn't want to confuse the user with the common parameters:

```
Param
(
    [Parameter(Mandatory)]
    [string]
    $Param1,

    # Dummy parameter to hide common parameters
    [Parameter(DontShow)]
    $Dummy
)
```

Just to be clear here, I'm obviously not arguing to keep the current behavior, I'm just saying that if we fix the current behavior we need to provide an alternative for the people that actually wants to hide common parameters.

**Comment by dwtaber** [comment]
> > Sure you can, but then you miss out on pipeline support, parametersets, and mandatory parameters.
> 
> Then you want an advanced function/script and not a simple function / script which again comes back to a more important question on function/script design. which is a much larger question than many realise until later on in the development of it & often changes drastically over time.

Sometimes one writes advanced functions for simple users.

I agree with @MartinGC94 that there's value in separating `DontShow` from hiding common parameters and keeping both capabilities.  Personally, for the use-case of making things less overwhelming for less skilled users, I think I'd rather see that handled by something like a `$CommonParameterVisibility` preference variable, rather than having to deal with it on a per-function basis.

**Comment by theJasonHelmick** [comment]
The Working Group has reviewed this -- we will investigate the original decision. The general feeling is that a parameter attribute should not affect other parameters or the common parameters. However, we would like to investigate the original reasoning behind disabling common parameters before making any changes - as it is very clear it was the desired behavior at the time.

**Comment by MartinGC94** [comment]
@theJasonHelmick Any update from the working group?

**Comment by Float-Moon** [comment]
I agree that there should be two separate attribute parameters, `DontShow` and something like MartinGC94 proposed, `HideCommonParameters`.

> TLDR: 
> As a part of `parameter attributes`, `DontShow` should let PowerShell automatically adds the Common Parameters, just like other `parameter attirbutes`
> 
> after define `[Parameter(DontShow)]`, there is no way to inform user in tab completion, that they can choose `$Verbose` param (or any other common parameters).
>
> And then we can not use custom `verbose` parameter for our logic,  although it's a common parameter in many terminal applications.

Here are details:

1. sometimes we want to use `Parameter Attribute` but provides concise tab completion options for users, as MartinGC94 metioned in https://github.com/PowerShell/PowerShell/issues/18472#issuecomment-1311069787

1. another scenario is making `$Verbose`  works properly with `[Parameter(DontShow)]`

```pwsh
function NeedVerboseParam {
    param(
        [ValidateSet("All", "Windows", "Linux", "MacOS")]
        [string]
        $Platform
        ,
        # need this param to tell wheter this function is called from itself
        [Parameter(DontShow)]
        [switch]
        $__CallFromItself
    )

   Write-Verbose "Choose platform $Platform..."


    if ($Platform -eq "All") {
        # spearate calls for each platform
        # so that no independent logic for [-Platform All]
        # which makes code cleaner
        NeedVerboseParam -Verbose $Verbose -Platform "Windows" -__CallFromItself $__CallFromItself
        NeedVerboseParam -Verbose $Verbose -Platform "Linux" -__CallFromItself $__CallFromItself
        NeedVerboseParam -Verbose $Verbose -Platform "MacOS" -__CallFromItself $__CallFromItself
        return
    }

    # platform specified codes....
}

```

- When user use this script, they are not aware there is verbose output as their tab completions not showing `$Verbose` options
```pwsh
# we can call verbose although it's not in tab completion
$ NeedVerboseParam -Verbose
VERBOSE:  Choose platform $Platform...
```

- Neither can we define a custom `Verbose` here, as it was used for common parameters (that's why there is need of `HideCommonParameters`)
```pwsh
function NeedVerboseParam {
    param(
        # add a custom verbose
       [switch]
       $Verbose
        .....
       [Parameter(DontShow)]
        [switch]
        $__CallFromItself
    )

```
```pwsh
$ NeedVerboseParam
NeedVerboseParam: A parameter with the name 'Verbose' was defined multiple times for the command.
```

-  If not using "advanced functions" (that use the CmdletBinding attribute or the Parameter attribute), sure we can add custom verbose, but will miss a lot of advanced features.
